bson/string_codec.go

```diff
@@ -58,12 +57,10 @@ func (sc *stringCodec) decodeType(_ DecodeContext, vr ValueReader, t reflect.Typ
 		if err != nil {
 			return emptyValue, err
 		}
-		if sc.decodeObjectIDAsHex {
+		if sc.decodeObjectIDAsHex || dc.objectIDAsHexString {
 			str = oid.Hex()
 		} else {
-			// TODO(GODRIVER-2796): Return an error here instead of decoding to a garbled string.
-			byteArray := [12]byte(oid)
-			str = string(byteArray[:])
+			return emptyValue, fmt.Errorf("decoding an object ID to a hexadecimal string is disabled by default")
```

Comment:
I believe the original intent was for mgo compatibility. See an example of the mgo behavior [here](https://go.dev/play/p/rTw5HuE8MW0). That is supported by the context of PR https://github.com/mongodb/mongo-go-driver/pull/213, which added the code. However, it's not clear why that behavior was supported by default, either in mgo or in the Go Driver. My best guess is that it was unintentional and was an oversight in both cases because the result of interpreting object ID bytes as a string is almost always invalid UTF-8, which could cause issues with lots of systems.